diff --git a/sql/wsrep_hton.cc b/sql/wsrep_hton.cc
index c096d26..27026ba 100644
--- a/sql/wsrep_hton.cc
+++ b/sql/wsrep_hton.cc
@@ -22,6 +22,7 @@
 #include "wsrep_xid.h"
 #include <cstdio>
 #include <cstdlib>
+#include "sql_parse.h" // sql_command_flags[]
 
 extern ulonglong thd_to_trx_id(THD *thd);
 
@@ -126,7 +127,8 @@ void wsrep_post_commit(THD* thd, bool all)
      /*
        Non-InnoDB statements may have populated events in stmt cache => cleanup
      */
-     WSREP_DEBUG("cleanup transaction for LOCAL_STATE: %s", thd->query());
+     if (sql_command_flags[thd->lex->sql_command] & CF_CHANGES_DATA)
+       WSREP_DEBUG("cleanup transaction for LOCAL_STATE: %s", thd->query());
      wsrep_cleanup_transaction(thd);
      break;
    }
